kpi-tracker/src/main.rs

```diff
@@ -0,0 +1,229 @@
+use std::collections::{BTreeMap, HashMap};
+use std::hash::Hash;
+
+use anyhow::Context;
+use chrono::{DateTime, Utc};
+use clap::Parser;
+use concordium_rust_sdk::{
+    id::types::AccountCredentialWithoutProofs,
+    smart_contracts::common::AccountAddress,
+    types::{
+        hashes::{BlockMarker, HashBytes},
+        AbsoluteBlockHeight,
+    },
+    v2::{self, AccountIdentifier, Client, Endpoint, FinalizedBlockInfo},
+};
+use futures::{self, StreamExt};
+
+#[derive(Debug, Parser)]
+struct Args {
+    /// The node used for querying
+    #[arg(long, default_value = "http://localhost:20001")]
+    node: Endpoint,
+    /// How many blocks to process.
+    // Only here for testing purposes...
+    #[arg(long, default_value_t = 10000)]
+    num_blocks: u64,
+}
+
+#[derive(Debug)]
+struct AccountDetails {
+    is_initial: bool,
+    block_hash: String, // FK to blocks
+}
+
+// Blocks are stored, so other tables can reference information about the block they were created in.
+#[derive(Debug)]
+struct BlockDetails {
+    block_time: DateTime<Utc>,
+    height: u64, // Used as a reference from where to restart on service restart.
+}
+
+type AccountsTable = HashMap<CanonicalAccountAddress, AccountDetails>;
+type BlocksTable = HashMap<String, BlockDetails>;
+
+struct DB {
+    blocks: BlocksTable,
+    accounts: AccountsTable,
+}
+
+#[derive(Eq, Debug, Clone, Copy, Ord, PartialOrd)]
+struct CanonicalAccountAddress(AccountAddress);
+
+impl From<CanonicalAccountAddress> for AccountAddress {
+    fn from(caa: CanonicalAccountAddress) -> Self {
+        caa.0
+    }
+}
+
+impl PartialEq for CanonicalAccountAddress {
+    fn eq(&self, other: &Self) -> bool {
+        let bytes_1: &[u8; 32] = self.0.as_ref();
+        let bytes_2: &[u8; 32] = other.0.as_ref();
+        bytes_1[0..29] == bytes_2[0..29]
+    }
+}
+
+impl Hash for CanonicalAccountAddress {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let bytes: &[u8; 32] = self.0.as_ref();
+        bytes[0..29].hash(state);
+    }
+}
+
+impl AsRef<CanonicalAccountAddress> for AccountAddress {
+    fn as_ref(&self) -> &CanonicalAccountAddress {
+        unsafe { std::mem::transmute(self) }
+    }
+}
+
+/// Returns a Map of AccountAddress, AccountDetails pairs not already included in `accounts_table`
+async fn new_accounts_in_block(
+    node: &mut Client,
+    block_hash: &HashBytes<BlockMarker>,
+    accounts_table: &AccountsTable,
+) -> anyhow::Result<BTreeMap<CanonicalAccountAddress, AccountDetails>> {
+    let mut accounts = node
+        .get_account_list(block_hash)
+        .await
+        .context(format!("Could not get accounts for block: {}", block_hash))?
+        .response;
+
+    let mut new_accounts = BTreeMap::new();
+
+    while let Some(res) = accounts.next().await {
+        let account = res.context("What exactly is this status error?")?; // TODO
```

Comment:
It could be that part of the stream could not be sent. This should be handled as a hard error and we fail in the same way as if we fail to get the initial stream.